### DISCUSSION TITLE
4.0

### DISCUSSION THREAD

**Discussion by ronag** [nodejs/undici#661] [General]
I have created a 4.0 milestone and added issues I believe would be good to sort before release.

Please take a look and let me know if you are missing something or disagree with any of the existing items.

**Comment by ronag** [comment]
@nodejs/undici

**Comment by ronag** [comment]
I've done some initial issue assignment. If anyone wants to grab any of the issues (or don't want them) feel free to re-assign or comment.

**Comment by ronag** [comment]
Updated milestone with what I believe is left https://github.com/nodejs/undici/milestone/4

**Comment by Ethan-Arrowood** [comment]
Converting this to a discussion to keep issue list directly related to features and fixes

**Comment by ronag** [comment]
Blockers for v4 release:

- https://github.com/nodejs/undici/issues/659
- https://github.com/nodejs/undici/issues/650

**Comment by ronag** [comment]
@nodejs/undici 

We are getting close to a v4 release. I don't see any further changes required in the code. We have some docs, typedefs and coverage fixes remaining. Is there anything else we need prior to a release? I will try to sort out https://github.com/nodejs/undici/pull/698 but I think we could easily do a semver minor release with that later.

How should we do the release? As usual?

Do we want to make a blog post or something (any volunteer)? 

Do we need a migrating section in the docs or is it just enough to document semver changes in the release notes? I don't think most users will be affected by the semver major changes.

**Reply by mcollina** [reply]
> How should we do the release? As usual?

Yes

> Do we want to make a blog post or something (any volunteer)?

I would recommend we do a blog post, ideally on https://nodejs.org/en/blog/ - https://github.com/nodejs/nodejs.org/tree/master/locale/en/blog.

We can probably coordinate this on a google doc. I will also ping the OpenJS Foundation marketing team for resharing etc.

@bnb wdyt?

> Do we need a migrating section in the docs or is it just enough to document semver changes in the release notes? I don't think most users will be affected by the semver major changes.

All my code broke because I was using `Agent.get()`. I think that's needed.

**Reply by ronag** [reply]
Added https://github.com/nodejs/undici/issues/707

**Reply by ronag** [reply]
https://github.com/nodejs/undici/issues/708

@mcollina Would you (or someone) mind setting up the google doc and create a skeleton? I'm more than happy to fill out details but would appreciate some help with the layout and body text and maybe some examples of how it could look (would be first time for me).

I think since this is the first blog post for undici some more general introduction and background for undici might be appropriate to include in the beginning.

**Reply by mcollina** [reply]
I'll do that.

**Comment by Ethan-Arrowood** [comment]
Can we do a preview v4 release and ask for early feedback? Or are we just going to cut the major and then start iterating on minor/patch versions from there?

**Reply by ronag** [reply]
I'm fine with either option.

**Comment by ronag** [comment]
I've made a 4.0.0-alpha.0 pre-release and tried to create somewhat of a change log https://github.com/nodejs/undici/releases/tag/v4.0.0-alpha.0.

@nodejs/undici feedback welcome!

**Comment by ronag** [comment]
So I think we are pretty much ready for a 4.0 release. At least code wise. I've updated the changelog here (https://github.com/nodejs/undici/releases/tag/v4.0.0-alpha.0).

The intention of doing the blogpost and breaking change guide might have been a bit overly ambitious of me.

**Reply by mcollina** [reply]
I think we should release a release candidate `-rc.1` while we work on the announcement.

**Reply by ronag** [reply]
Should we publish current master as -rc.1?

**Reply by mcollina** [reply]
Yes but not as latest.

**Reply by ronag** [reply]
Done
